### DISCUSSION TITLE
Contains Eager not working properly when query executes with `.first()` but works as expected when query executes with `.all()`

### DISCUSSION THREAD

**Discussion by faizanalhassan** [sqlalchemy/sqlalchemy#11982] [Usage Questions] [answered]
### Describe the bug

I have two ORM classes, `Product` and `UnitPrice`, structured as follows:

```python
class Product(db.Model):
    __tablename__ = "products"
    ...
    unit_prices = relationship("UnitPrice")
    ...
    
class UnitPrice(db.Model):
    __tablename__ = "unit_prices"
    ...
    product_id = sa.Column(
        sa.Integer,
        sa.ForeignKey("products.id", ondelete="CASCADE"),
        nullable=False,
    )
    ...
```

When querying these models, I encountered an inconsistency between using `.first()` and `.all()` while fetching related `UnitPrice` entries for a `Product`.

### Query Behavior

1. **Using `.first()`**:

```python
Product.query.join(Product.unit_prices)
    .options(sa_orm.contains_eager(Product.unit_prices))
    .filter(Product.id == 7472, UnitPrice.channel_id == 4)
    .first().unit_prices
```

This returns a list with **only one result**: `[<UnitPrice 47162>]`.

2. **Using `.all()`**:

```python
Product.query.join(Product.unit_prices)
    .options(sa_orm.contains_eager(Product.unit_prices))
    .filter(Product.id == 7472, UnitPrice.channel_id == 4)
    .all()[0].unit_prices
```

This returns the **expected result with two unit prices**: `[<UnitPrice 47162>, <UnitPrice 47163>]`.

### Issue

I expect both queries to return consistent results. In this case, `.first()` is returning only one `UnitPrice` object, while `.all()` returns two (which is correct). It seems that when calling `.first()`, one of the associated `UnitPrice` records is being excluded.

### Expected Behavior

Both `.first()` and `.all()` should return the same list of `UnitPrice` objects when querying the `unit_prices` relationship for the same `Product`.

### Additional Information

- SQLAlchemy version: 2.0.20
- Database: PostgreSQL 16

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#sqlalchemy.orm.contains_eager

### SQLAlchemy Version in Use

2.0.20

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 16

### Python Version

3.10

### Operating system

OSX

### To Reproduce

```python
class Product(db.Model):
    __tablename__ = "products"
    ...
    unit_prices = relationship("UnitPrice")
    ...
    
class UnitPrice(db.Model):
    __tablename__ = "unit_prices"
    ...
    product_id = sa.Column(
        sa.Integer,
        sa.ForeignKey("products.id", ondelete="CASCADE"),
        nullable=False,
    )
    ...
assert len(Product.query.join(Product.unit_prices)
    .options(sa_orm.contains_eager(Product.unit_prices))
    .filter(Product.id == 7472, UnitPrice.channel_id == 4)
    .all()[0].unit_prices) > 1

assert len(Product.query.join(Product.unit_prices)
    .options(sa_orm.contains_eager(Product.unit_prices))
    .filter(Product.id == 7472, UnitPrice.channel_id == 4)
    .first().unit_prices) > 1
```

### Error

```
# Copy the complete stack trace and error message here, including SQL log output if applicable.
```

**Comment by CaselIT** [comment]
Hi,

This is expected, since first applies a limit 1 to the sql.
Since you are filtering by primary key you may want to use https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.one or https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.one_or_none instead

**Comment by faizanalhassan** [comment]
@CaselIT `joinedload` works fine with `.first()`, so I don't think this is the correct reason. I strongly believe there is some bug. 
`Product.query.options(sa_orm.joinedload(Product.unit_prices)).filter(Product.id==7472).first().unit_prices
` gives me 4 results.
if applying `.first()` was applying limit it should apply on all loading techniques. But seems like the issue is only with `contains_eager`.

**Reply by CaselIT** [reply]
joinedload has special handling for this use case (that I'm guessing it's not great cc @zzzeek ).
What you are asking here is that `first()` takes your join query and disregards that, and instead creates a subquery from the primary from with limit applied, then joins that subquery.
That would be hugely surprising to people.
Also should it do that only if contains eager is applied? That would be even stranger

In any case the fact that query tried to do clever things with first leading to poor result it's the reason why it's legacy (ake feature freazed). So a change in behaviour here is hugely unlikely

**Comment by zzzeek** [comment, marked as answer]
legacy `Query.first()` applies LIMIT 1 to the SQL.  joinedload() detects this and produces a subquery for the original filtered condition with the LIMIT then joinedloads around that.  if you use contains_eager(), you're writing the SQL completely, so if you want LIMIT 1 in your query, you'd have to apply subqueries appropriately.   the use of contains_eager() assumes nothing about the structure of the SQL so no editorial decisions like this are taken (if you want the editorial decisions, just use joinedload()).

Easiest solution here is to not apply LIMIT 1.   Using modern `select()` with `session.scalars()` will give you a `first()` method that does not suddenly apply LIMIT unexpectedly.   See https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#orm-query-unified-with-core-select

**Reply by faizanalhassan** [reply]
Thanks for the explanation. I am clear about the reason now.
I am more comfortable with `.one` suggested by @CaselIT, which requires me less code changes.

**Reply by CaselIT** [reply]
> joinedload() detects this and produces a subquery for the original filtered condition with the LIMIT then joinedloads around that.

does that work like that also when using a select?

**Reply by zzzeek** [reply]
yes it does.   the point is LIMIT isn't added unless you actually say `limit()`
